### ISSUE TITLE
Authentication and Authorization

### ISSUE THREAD

**Issue by markrey** [liftbridge-io/liftbridge#36]
Hi,

I see that Authentication & Authorization is in planned features.
Any specific plans for using embedded DB / file based or others? Any reference architecture would help.

Thanks

**Comment by hobti01** [comment]
Why would this be different from nats [authentication](https://nats.io/documentation/server/gnatsd-authentication/) and [authorization](https://nats.io/documentation/server/gnatsd-authorization/)?

**Comment by tylertreat** [comment]
It's different in that Liftbridge is a consumer of NATS, storing messages from it, and making those messages available to Liftbridge consumers. The way that is implemented in practice could absolutely work like the way NATS does however.

**Comment by oliverhausler** [comment]
IMO username/password auth will not work in a distributed world, especially as credentials change over time, and sync and lookup is required. Instead, I'd suggest using JWTs:

- The server is statically configured with a single secret (which all servers in the cluster know).
- Client ships a signed JWT with a claim containing permissions what the client can do (subjects, etc.).
- If the signature in the JWT is authentic, server serves the client with that token until token expires.

This lets a management server create tokens, which Liftbridge can use independent of any outside systems or databases.

**Comment by paddybyers** [comment]
I think that there could legitimately be multiple ways to support authentication and authorisation, especially where you want liftbridge to enforce access control at the stream level.

However, we have a need just to be able to authenticate client connections without any fine-grained access control and propose supporting client certificate authentication as an option. @tylertreat would you welcome a PR where we add configuration option(s) to enable that (including also any support needed in the client library) ?

**Comment by tylertreat** [comment]
@paddybyers Yes, I think client certificate authentication is a good idea and would love a PR.

**Comment by LaPetiteSouris** [comment]
@tylertreat I am taking a look into this issue. It seems that authentication of client via TLS has already been in place ?

This piece of code on server already takes care the TLS authentication of grpc client:  https://github.com/liftbridge-io/liftbridge/blob/c0c3e83d853abc728f7cd55618a062292107f0bc/server/server.go#L442

The only thing missing is may be a dedicated section on the documentation to document this behavior.

The authorization part is not available though. I am looking into an Access Control List (ACL) approach on stream level.

I intend to pick ACL for few reasons:

- We may want to give permissions up to stream level, so RBAC seems ill-suited
- ACL is battle-proven for streaming system (Kafka, Pulsar).

I am more in favor of using [casbin](https://github.com/casbin/casbin) with a SQL connection (supported also via [Casbin SQL Adapter](https://casbin.org/docs/en/adapters) to store users information.

Any thought ?

**Comment by tylertreat** [comment]
@LaPetiteSouris Yes, TLS client auth was implemented since the discussion above. This issue was never updated. I think a dedicated section in the docs would be good. We probably ought to add an Authentication page (and include Authorization once supported).

I think ACL is probably the right choice for authorization. I have not used Casbin myself, but I am aware of it and it seems to be the most mature option for Go. I'm apprehensive about requiring a SQL database for authorization since I would prefer to keep external dependencies minimal. Note that Liftbridge relies on BoltDB (etcd's [bbolt](https://github.com/etcd-io/bbolt) fork actually) internally for its Raft store, which is an embedded kv database. It does look like there is a [Casbin adapter for Bolt](https://github.com/speza/casbin-bolt-adapter), but it appears to be experimental. Not saying we _have_ to use Bolt, but it is something we should consider. IMO we should think carefully about what the backend implementation should look like. I would prefer to not introduce a new external dependency if at all possible, however.
